### ISSUE TITLE
typemapping c++ boost::optional<int> to C# int?

### ISSUE THREAD

**Issue by nanan14** [swig/swig#1307]
Hi ,

is there a way to typemap boost::optional<int> to c# int? and vice versa
and also for other type like bool  , string and other classes?

so far I only found this implementation.
https://stackoverflow.com/questions/15912937/can-you-swig-a-boostoptional
```
%template(optionalInt) optional<int>;
%typemap(ctype) optional<int> "void *"; 
%typemap(imtype) optional<int> "global::System.IntPtr"; 
%typemap(cstype) optional<int> "int?"; 
%typemap(csin) optional<int> "$csinput";
%typemap(out) optional<int> %{
    std::vector<int> result_vec;
    if (!!$1)
    {
        result_vec = std::vector<int>(0, $1.get());
    }
    else
    {
        result_vec = std::vector<int>();
    }
    $result = new std::vector< int >((const std::vector< int > &)result_vec); 
%}
%typemap(csout, excode=SWIGEXCODE) boost::optional<int> {
     global::System.IntPtr cPtr =$imcall;$excode
	 if (cPtr != global::System.IntPtr.Zero)
	 {
	     return (int)($imcall);
	 }
	 return null;
}
%typemap(csvarout, excode=SWIGEXCODE) optional<int>
%{ 
    get 
    { 
         global::System.IntPtr cPtr =$imcall;$excode
	    if (cPtr != global::System.IntPtr.Zero)
	    {
	        return (int)($imcall);
	    }
	    return null;
    } 
%}
```

**Comment by vadz** [comment]
There are no standard typemaps for this in SWIG but I do have my own typemaps for `std::optional`-like class for C#, Java and Python (OK, so the latter one is trivial, but still). I could tidy them up for the inclusion in SWIG itself if there is sufficient interest.

**Comment by aksswami** [comment]
It would be great if you can share the snippet as I am also looking for the same. I  am new to the SWIG and type mapping complex type is something I don’t understand as of now.

**Comment by wsfulton** [comment]
@vadz, yes! They ought to port to `std::optional` which we ought to also have in SWIG.

**Comment by diegoferigo** [comment]
@vadz Is the `std::optional` support a feature you are still working on? Can you please provide the snippet you mentioned in the meantime?

**Comment by vadz** [comment]
No, sorry, other higher priority tasks took over and I was never able to work on this.

I've created some gists showing the typemaps we're using:

1. [Main file](https://gist.github.com/vadz/1d7a7c1be4f55fe45b46ea06454349e1) including language-specific typemaps definitions.
1. [C# typemaps](https://gist.github.com/vadz/c6a9cf0c62d8db7ad687992735a9ae08) (you may not need VBA-specific stuff)
1. [Java typemaps](https://gist.github.com/vadz/7ba2bfb73d04483e2f6254b05feb7e1f) (requires Java 8)
1. [Python typemaps](https://gist.github.com/vadz/44bcb03ea2ca8a370012a1f47aa00eaf)

These typemaps are for a very simple custom `OptionalValue<>` type, but should be straightforward to adapt for `std::optional<>`.

**Comment by diegoferigo** [comment]
Thanks a lot @vadz, your files have been very helpful. I am currently using a very simple swig configuration for `std::optional` using this [`optional.i`](https://github.com/robotology/gym-ignition/blob/master/bindings/optional.i). It requires to instantiate templates with something similar to the following:

```
%include "optional.i"
%template(Optional_i) std::optional<int>;
```

---

I tried to make something more complex, similar to a templated typemap in order to have a more idiomatic usage from python, but I couldn't make it work. For instance, I would like to directly map optional numeric types and strings to the target language. In python:

```python
myint = mymodule.Optional_i(42)
type(myint) # <class 'int'>

myint = mymodule.Optional_i()
type(myint) # <class 'NoneType'>
```


I am pretty sure that it can be done using templated `fragments` together with `$typemap(out,T)` in language-specific typemaps, but I don't have enough experience with swig.

Does anyone have any example? I tried to understand how is done for `std::vector` but I didn't get it entirely.

**Comment by vadz** [comment]
> In your example I have't understood the logic behind `DEFINE_OPTIONAL_CLASS` and `DEFINE_OPTIONAL_SIMPLE`.

In C# the representation of C++ `optional<T>` is either `T?` if it's a simple class or just `T` if it's a class. A single macro wouldn't be able to distinguish between these 2 cases (or, at least, I don't know how would it do it), so I've created 2 of them.

> Do you have to instantiate your templates with `%template`?

No, the macros contain it.

For idiomatic usage, you really shouldn't be using `Optional_i` at all. Just use integers or `None` on the Python side.

**Comment by diegoferigo** [comment]
> In C# the representation of C++ optional<T> is either T? if it's a simple class or just T if it's a class. A single macro wouldn't be able to distinguish between these 2 cases (or, at least, I don't know how would it do it), so I've created 2 of them.

I'm not really familiar with C#, but I think I got your point.

> For idiomatic usage, you really shouldn't be using Optional_i at all. Just use integers or None on the Python side.

I updated my comment above to reflect what I really meant. Now it should be more clear.

**Comment by JacobVarghese1992** [comment]
Hey @vadz , Sorry I'm looking at your java version. What exactly is OptionalValue in there. I get a Template 'OptionalValue' is undefined error.

**Comment by vadz** [comment]
`OptionalValue` is our custom type similar to `boost::optional` and `std::optional`, you should replace it with the type you actually use for your optional values.

**Comment by SmrtSmrtnik** [comment]
+1 for feature request, for what it's worth.
i have lots of c++ code with std::optional all over the place I've been tasked to SWIGify into Java, so definitely going to be looking at @vadz  link and https://stackoverflow.com/questions/30272562/swig-wrap-boostoptional-in-java + the JNI spec as I've never done JNI before (and new to SWIG). Good times.

**Comment by lluisalemanypuig** [comment]
+1 for feature request. I am also very interested in seeing this implemented in SWIG.
